### ISSUE TITLE
NullPointerException during WrappedSession.removeAttribute(String)

### ISSUE THREAD

**Issue by TobseF** [vaadin/multiplatform-runtime#98]
### Describe the bug

I noticed a `NullPointerException` during a `handleRequest` beause 
> _`Cannot invoke "com.vaadin.flow.server.WrappedSession.removeAttribute(String)" because "httpSession" is null`_

Because the stacktrace is triggered by a request and not during page load or a special site interaction, it's hard to reproduce it. Anyway the question is, if this `removeAttribute(String)` function shouldn't be null-safe. Or maybe one step further, the `WrappedSession` should not be created, with a nulled Session. The exception happens in Vaadin internals, so it's not clear if I should try to catch this Exception or if this is a framework bug.

The setup uses the [MPR](https://vaadin.com/multiplatform-runtime), but the stracktrace seems to bee independent from the MPR stuff.

Maybe the bug is a also related to: https://github.com/vaadin/multiplatform-runtime/issues/95

### ⚡ Stacktrace
``` java
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.WrappedSession.removeAttribute(String)" because "httpSession" is null
caused by: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.WrappedSession.removeAttribute(String)" because "httpSession" is null
at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
at com.vaadin.flow.server.FutureAccess.get(FutureAccess.java:62)
at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2056)
at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:699)
at com.vaadin.flow.server.VaadinService.requestEnd(VaadinService.java:1498)
at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1573)
at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:252)
at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:111)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:584)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
```

### System Info

Project with MPR
vaadin-14.version: `14.7.1`
vaadin-8.version: `8.14.1`
vaadin.flow.plugin.version: `2.7.2`

**Comment by denis-anisimov** [comment]
Looks like duplicate of https://github.com/vaadin/multiplatform-runtime/issues/93

**Comment by ZheSun88** [comment]
the fix that Denis mentioned has been released with **MPR 2.2.5** which was included in **Vaadin 14.7.2** 
Can you try the latest version and let us know? @TobseF. Thank you

**Comment by TobseF** [comment]
I'm testing the **Vaadin 14.7.2** release. First I found another session related exception which seems to very analog. I opened a new ticket for that one:
[NullPointerException in VaadinService.requestEnd because the return value of "VaadinRequest.getAttribute(s)" is null ](https://github.com/vaadin/flow/issues/12099)

Then I checked the original test for the problem in this thread. And it still happens in the same way:
### ℹ System Info

vaadin-14.version: `14.7.2`
vaadin-8.version: `8.14.1`
mpr-core: `2.2.5`
vaadin.flow.plugin.version: `2.7.2`

### ⚡ Stacktrace
``` java
2021-10-20 11:34:37,530 ERROR --- [1567202570-1345] c.v.flow.server.DefaultErrorHandler      :  [requestContext=Update UI, requestId=974, proxyRefId=007b02dd-5e74-436f-bb7c-48558cff4081, domain=localhost, testLogID=XfrH7_de.itscope.redgiant.testbench.tests.quote.Quote_ProjectPrice, requestUri=/red/, RequestType=UIDL] 

java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.WrappedSession.removeAttribute(String)" because "httpSession" is null
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at com.vaadin.flow.server.FutureAccess.get(FutureAccess.java:62)
	at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2056)
	at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:713)
	at com.vaadin.flow.server.VaadinService.requestEnd(VaadinService.java:1498)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1573)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:252)
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:111)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:584)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at de.itscope.mv.service.order.receiver.OrderReceiverAuthFilter.doFilter(OrderReceiverAuthFilter.java:62)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:209)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at de.itscope.servlet.filter.RequestTraceFilter.doFilterInternal(RequestTraceFilter.java:80)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:109)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at de.itscope.servlet.filter.RequestMetricFilter.doFilterInternal(RequestMetricFilter.java:105)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:590)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.__doHandle(ContextHandler.java:1300)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:42020)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.NullPointerException: Cannot invoke "com.vaadin.flow.server.WrappedSession.removeAttribute(String)" because "httpSession" is null
	at com.vaadin.mpr.core.AbstractLegacyWrapper.handleDetach(AbstractLegacyWrapper.java:168)
	at com.vaadin.flow.component.ComponentEventBus.fireEventForListener(ComponentEventBus.java:205)
	at com.vaadin.flow.component.ComponentEventBus.fireEvent(ComponentEventBus.java:194)
	at com.vaadin.flow.component.Component.fireEvent(Component.java:358)
	at com.vaadin.flow.component.ComponentUtil.fireEvent(ComponentUtil.java:386)
	at com.vaadin.flow.component.ComponentUtil.onComponentDetach(ComponentUtil.java:264)
	at java.base/java.util.Optional.ifPresent(Optional.java:176)
	at com.vaadin.flow.internal.nodefeature.ComponentMapping.onDetach(ComponentMapping.java:109)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.vaadin.flow.internal.StateNode.forEachFeature(StateNode.java:351)
	at com.vaadin.flow.internal.StateNode.fireDetachListeners(StateNode.java:837)
	at com.vaadin.flow.internal.StateNode.onDetach(StateNode.java:341)
	at com.vaadin.flow.internal.StateNode.setParent(StateNode.java:280)
	at com.vaadin.flow.internal.StateTree$RootNode.setParent(StateTree.java:65)
	at com.vaadin.flow.component.internal.UIInternals.setSession(UIInternals.java:401)
	at com.vaadin.flow.server.VaadinSession.removeUI(VaadinSession.java:623)
	at com.vaadin.flow.server.VaadinService.lambda$null$65d65655$1(VaadinService.java:665)
	at com.vaadin.flow.component.UI.accessSynchronously(UI.java:434)
	at com.vaadin.flow.component.UI.accessSynchronously(UI.java:395)
	at com.vaadin.flow.server.VaadinService.lambda$fireSessionDestroy$9c853e43$1(VaadinService.java:655)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2053)
	... 67 common frames omitted
````

### 🙏 So please reopen the issue for further investigations.

**Comment by denis-anisimov** [comment]
Oh, sorry, my bad.
It's not a duplicate.
The #93 is about ISE, this one is NPE.
I don't know what has happened to me when I closed it.

Reopened but this has chances to be fixed if you provide the steps to reproduce.

**Comment by TatuLund** [comment]
> at com.vaadin.mpr.core.AbstractLegacyWrapper.handleDetach(AbstractLegacyWrapper.java:168)

I have seen this reported by another user to me. I have tried to reproduce the problem in simplified application, but so far I have not been able to. So clearly it needs some very specific scenario.

Also I have seen reports 2-3 other cases where session disappears mysteriously, but the NPE happens in different part of the code flow.

Something goes wrong somewhere, but why, that I do not know yet.

Here is a test app that I have used to recreate the problem (as mentioned, the problem is not happening there) https://github.com/TatuLund/mpr-demo/tree/mpr-4
